src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteCache.java

```diff
@@ -248,50 +256,149 @@ public void ensureInputsPresent(
 
-    @Nullable
-    HashingOutputStream hashOut =
-        options.remoteVerifyDownloads ? digestUtil.newHashingOutputStream(out) : null;
+    @Nullable Supplier<HashCode> hashSupplier = null;
+    if (options.remoteVerifyDownloads) {
+      HashingOutputStream hashOut = digestUtil.newHashingOutputStream(out);
+      hashSupplier = hashOut::hash;
+      out = hashOut;
+    }
+
     SettableFuture<Void> outerF = SettableFuture.create();
+    Futures.addCallback(
+        downloadBlob(resourceName, digest, out, hashSupplier),
+        new FutureCallback<Void>() {
+          @Override
+          public void onSuccess(Void result) {
+            outerF.set(null);
+          }
+
+          @Override
+          public void onFailure(Throwable t) {
+            outerF.setException(t);
+          }
+        },
+        Context.current().fixedContextExecutor(MoreExecutors.directExecutor()));
+    return outerF;
+  }
+
+  static class ProgressiveBackoff implements Backoff {
+    private final Supplier<Backoff> backoffSupplier;
+    private Backoff currentBackoff = null;
+    private int retries = 0;
+
+    /**
+     * Creates a resettable Backoff for progressive reads.
+     * After a reset, the nextDelay returned indicates an immediate retry.
+     * Initially and after indicating an immediate retry, a delegate is generated
+     * to provide nextDelay until reset.
+     *
+     * @param backoffSupplier Delegate Backoff generator
+     */
+    ProgressiveBackoff(Supplier<Backoff> backoffSupplier) {
+      this.backoffSupplier = backoffSupplier;
+      currentBackoff = backoffSupplier.get();
+    }
+
+    public void reset() {
+      if (currentBackoff != null) {
+        retries += currentBackoff.getRetryAttempts();
+      }
+      currentBackoff = null;
+    }
+
+    @Override
+    public long nextDelayMillis() {
+      if (currentBackoff == null) {
+        currentBackoff = backoffSupplier.get();
+        retries++;
+        return 0;
+      }
+      return currentBackoff.nextDelayMillis();
+    }
+
+    @Override
+    public int getRetryAttempts() {
+      int retryAttempts = retries;
+      if (currentBackoff != null) {
+        retryAttempts += currentBackoff.getRetryAttempts();
+      }
+      return retryAttempts;
+    }
+  }
+
+  private ListenableFuture<Void> downloadBlob(
+      String resourceName,
+      Digest digest,
+      OutputStream out,
+      @Nullable Supplier<HashCode> hashSupplier) {
+    Context ctx = Context.current();
+    AtomicLong offset = new AtomicLong(0);
+    ProgressiveBackoff progressiveBackoff = new ProgressiveBackoff(retrier::newBackoff);
+    return retrier.executeAsync(
+        () -> ctx.call(
+            () -> requestRead(
+                resourceName,
+                offset,
+                progressiveBackoff,
+                digest,
+                out,
+                hashSupplier)));
+  }
+
+  private ListenableFuture<Void> requestRead(
+      String resourceName,
+      AtomicLong offset,
```

Comment:
While I appreciate your ingenuity, I'd argue that this is quite the hack. I am wondering if this whole logic couldn't be implemented in the Retrier itself i.e. 

```
class ProgressiveRetrier extends Retrier {

  public ListenableFuture<Void> executeAsync(AsyncCallable<Long> call, long totalBytes) {
     return executeAsync(new AsyncCallable<Void>() {
       // implement progressive retry logic in here. in case of deadline exceeded and having made progress
       // immediately retry.
     }, newBackoff());
  }
}
```

I haven't fully thought it through but I think this could be made generic enough so that it could also be used in the `ByteStreamUploader` and the `HttpBlobStore`. Wdyt?

Reply:
whatever gets me to an accept at this point.

Reply:
Assuming that this comment addresses the progressive offset holder, I don't see how totalBytes can be used in the retrier to effect this: `call` presents an offset-worthy result only in the case of non-exception, but DEADLINE_EXCEEDED will be an exception state for the future. Do you expect the call to swallow the DEADLINE_EXCEEDED and respond with a committedSize, expecting the retrier to recall until it matches totalBytes? If so, the call still needs to maintain the current offset (per the AtomicLong), since there's no way to pass it back to the subsequent one.